libraries/botbuilder-core/botbuilder/core/activity_handler.py

```diff
@@ -344,3 +352,56 @@ class ActivityHandler:
             conversation update, message reaction, or event activity, it calls this method.
         """
         return
+
+    async def on_invoke_activity(  # pylint: disable=unused-argument
+        self, turn_context: TurnContext
+    ) -> Union[InvokeResponse, None]:
+        """
+        Registers an activity event handler for the _invoke_ event, emitted for every incoming event activity.
+
+        :param turn_context: The context object for this turn
+        :type turn_context: :class:`botbuilder.core.TurnContext`
+
+        :returns: A task that represents the work queued to execute
+        """
+
+        if (
+            turn_context.activity.name == SignInConstants.verify_state_operation_name
+            or turn_context.activity.name
+            == SignInConstants.token_exchange_operation_name
+        ):
+            return await self.on_sign_in_invoke(turn_context)
+
+        return await self.on_invoke(turn_context)
+
+    async def on_sign_in_invoke(  # pylint: disable=unused-argument
+        self, turn_context: TurnContext
+    ):
+        """
+        Invoked when a signin/verifyState or signin/tokenExchange event is received when the base behavior of
+        on_invoke_activity(TurnContext{InvokeActivity}) is used.
+        If using an OAuthPrompt, override this method to forward this Activity"/ to the current dialog.
+        By default, this method does nothing.
+
+        :param turn_context: The context object for this turn
+        :type turn_context: :class:`botbuilder.core.TurnContext`
+
+        :returns: A task that represents the work queued to execute
+        """
+        return
+
+    async def on_invoke(  # pylint: disable=unused-argument
```

Comment:
We don't want this one -  the model is flatter for invoke - this looks like a copy of the event-activity pattern.

Specifically we want an overridable onInvokeActivity and it should return an invoke-response then we want code in the onTurn that takes that invoke-response, creates an invoke-response-activity and sends it. The code inline in onTurn should do this send conditionally - if the override has already done a send then don't do one here - refer to the code in C# (ActivityHandler) and JS (ActivityHandlerBase).

Next having introduced this onInvokeActivity method in ActivityHandler we can remove the now duplicate code in TeamsActivityHandler.

So teh change is basically to move the code up the class hieerarchy from TeamsActivityHandler into ActivityHandler

Reply:
I just pushed those changes.